lib/ansible/galaxy/collection.py

```diff
@@ -1054,6 +1168,54 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
     return dependency_map
 
 
+def _collections_from_scm(collection, requirement, b_temp_path, force, parent=None):
+    """
+    Returns a list of collections found in the repo. If there is a galaxy.yml in the collection then just return
+    the specific collection. Otherwise, check each top-level directory for a galaxy.yml.
+
+    :param collection: URI to a git repo
+    :param requirement: The version of the artifact
+    :param b_temp_path: The temporary path to the archive of a collection
+    :param force: Whether to overwrite an existing collection or fail
+    :param parent: The name of the parent collection
+    :return: List of CollectionRequirement objects
+    """
+
+    reqs = []
+    name, version, path, fragment = parse_scm(collection, requirement)
+    b_repo_root = to_bytes(name, errors='surrogate_or_strict')
+
+    if fragment:
+        b_fragment = to_bytes(fragment, errors='surrogate_or_strict')
+        b_collection_path = os.path.join(b_temp_path, b_repo_root, b_fragment)
+    else:
+        b_collection_path = os.path.join(b_temp_path, b_repo_root)
+
+    b_galaxy_path = os.path.join(b_collection_path, b'galaxy.yml')
+
+    err = "{0} appears to be an SCM collection source, but the required galaxy.yml was not found. " \
+          "Append #path/to/collection/ to your URI (before the comma separated version, if one is specified) " \
+          "to point to a directory containing the galaxy.yml or directories of collections".format(collection)
+
+    display.vvvvv("Considering {0} as a possible path to a collection's galaxy.yml".format(b_galaxy_path))
+    if os.path.exists(b_galaxy_path):
+        return [CollectionRequirement.from_path(b_collection_path, force, parent, fallback_metadata=True)]
+
+    if not os.path.isdir(b_collection_path) or not os.listdir(b_collection_path):
+        raise AnsibleError(err)
+
+    for b_possible_collection in os.listdir(b_collection_path):
+        b_collection = os.path.join(b_collection_path, b_possible_collection)
+        b_galaxy = os.path.join(b_collection, b'galaxy.yml')
+        display.vvvvv("Considering {0} as a possible path to a collection's galaxy.yml".format(b_galaxy))
+        if os.path.exists(b_galaxy):
+            reqs.append(CollectionRequirement.from_path(b_collection, force, parent, fallback_metadata=True))
+        else:
+            raise AnsibleError(err)
```

Comment:
Should we only fail if `reqs` is an empty list, i.e. no `galaxy.yml` was found in any sub dirs? The way the docs are written is that this will only install dirs in SCM that contain the `galaxy.yml` one level deep. Right now it will fail unless either the root contains `galaxy.yml` or all the subdirs one level deep contain a `galaxy.yml`.

Reply:
I don't feel very strongly about this, but I did choose it intentionally. If someone points to a repo but nothing is found, it seems like it's most likely a mistake (similarly to how if a collection name doesn't exist on galaxy, trying to install it using the name is an error). Hopefully it will also make it easy to figure out fragments for a custom path. If people want to include "empty" looking repositories, they can use --ignore-errors like with any other installation errors. If you/others prefer failing only if reqs is an empty list I'll update that, but this made sense to me.